_posts/2023-2-10-profiling-with-signals.md

```diff
@@ -0,0 +1,257 @@
+---
+layout:     post
+title:      Profiling with signals
+date:       2023-2-10
+author:     Javier Honduvilla Coto
+categories: profiling signals
+---
+
+Sampling profilers periodically fetch the stacktrace of the application under observation. Historically, this has been from within the process, in-band,  using signals. There's one dedicated to this sole purpose, `SIGPROF`. We can register a signal handler for it that would retrieve the current stack. Then that stack could be aggregated in place or added to some global data structure and eventually, the generated profile can be written somewhere, such as on disk. Some care has to be taken but this is the general idea of how it works[0].
+
+## Advantages of signal-based profiling
+
+Using signals for profiling has several advantages. We can leverage the existing machinery from your runtime to do most of the heavy lifting. Many programming languages, such as Go or Python, have a standard library function to get the current stack[1]. There is no need to re-implement the stack unwinding mechanism, which is convenient because most languages, particularly high-level ones, don't have a stable stack ABI. We need to account for internal details that might change in future versions.
```

Comment:
Great point, thanks! 

For context (mostly for myself hehe) I was thinking of profilers implemented entirely in the target language, so without native extensions. I believe in Python and perhaps in other languages, execution within a signal handler has fewer foot-guns than in the C runtime. It would be very interesting to see how this is implemented (if I am not wrong!)